Breaking changes: improved shutdown and error approach, queue abstraction

jonsterling opened the pull request:
This is intended to address #19, #17, #21, and #22.

There is more to be done before this may be merged.
- [x] stop plumbing exceptions into `MonadError`
- [x] remove `either` dependency
- [x] expose possibly-throwing functions as returning `m (Either e a)`
- [x] use `enclosed-exceptions` where appropriate
- [x] implement graceful shutdown of producer 
- [x] add configurable timeout for producer worker cleanup
- [x] update changelog
- [x] remove support for `PutRecord` endpoint
- [x] implement new queue that resolves live-locking issue

jonsterling committed:
Breaking changes: improved shutdown and error approach

This is intended to address #19 and #17

There is more to be done before this may be merged.

jonsterling committed:
Remove exception wrappers

These are not desirable, since they destroy the canonicity of
exceptions.

jonsterling committed:
Version bump (0.4.0.0)

jonsterling committed:
Bring monad-control lower bound up to date

jonsterling committed:
[Producer] Add annoying type annotation to fix build error

jonsterling committed:
[Producer] name improvements

jonsterling committed:
Check in constraints file

jonsterling committed:
Cleanup: remove MonadKinesis; other improvemenets

- Remove `MonadKinesis`; it is easy enough just to pass the object
- Fix strange-but-probably-not-wrong error handling in runKinesis
- Remove dependency on `errors` package (and thence on `either` package)

jonsterling committed:
Clean up a few things using record wildcards

jonsterling committed:
[Consumer] export a few lenses that were forgotten

jonsterling committed:
[Consumer] some aesthetic cleanup

jonsterling committed:
[Producer] Add a timeout to the cleanup routine

jonsterling committed:
[Producer] Separate out some exceptions; fix throwing in writeProducer

jonsterling committed:
[Producer] More granular exceptions (reasons below)

The point of this is that these exceptions come from different places,
and it is not the case that in the course of handling one, you should be
expected to handle all the others. A common mistake would be to handle
one of them, and then get a pattern match failure because it was a
different one of the same type that got thrown.

jonsterling committed:
[Producer] remove managedBracket

jonsterling committed:
Updated CHANGELOG

jonsterling committed:
[Consumer] Remove unused ConsumerError type

jonsterling committed:
Remove @MonadBaseControl IO m@ where not necessary

jonsterling commented on src/Aws/Kinesis/Client/Producer.hs:
```diff
@@ -224,6 +224,10 @@ data ProducerKit
   -- ^ The number of requests to run concurrently (minimum: 1).
+
+  , _pkCleanupTimeout ∷ !(Maybe Int)
+  -- ^ The timeout in milliseconds, after which the producer's cleanup routine
```
`s/millisconds/microseconds`

jonsterling committed:
[Producer] Improve WriteProducerException

jonsterling committed:
[Producer] some refactoring

jonsterling committed:
Remove some class constraints in favor of simpler flow

- MonadKinesis
- MonadConsumerInternal
- MonadProducerInternal

It's a lot easier to just pass the objects explicitly.

jonsterling committed:
[Common] just put runKinesis directly in IO

jonsterling committed:
Put internal stuff directly in IO

These internal things can just be directly in IO; it actually makes
things a bit simpler. Exposed functions should be polymorphic in their
monad, but the internal stuff should be as simple as possible.

jonsterling committed:
[Producer] remove redundant annotation

jonsterling committed:
[Producer] rename _cpThrottlingDelay ⇒ _cpMinChunkingInterval

The original name made no sense

jonsterling committed:
[Producer] fix serious problems w/ chunking, cleanup and termination

I don't really know what I was thinking when I wrote the previous
version of chunking, which pretty much did not do what I wanted, and was
very convoluted.

Also, the new implementation of graceful shutdown was totally broken.

jonsterling commented:
@larskuhtz When you have a moment, would you mind taking a look at this pull request? It may be best to look mostly at the final diff, since things changed a bit along the way and each individual commit is perhaps not quite so informative.

The main changes here are as follows:
1. The code that consumed the input queue in "chunks" (in order to batch things in `PutRecords` requests) was pretty bizarre and wrong; I don't think that it would have caused any bugs, but it probably didn't behave quite how I had hoped. I have replaced it with code that seems a lot more straightforward and actually appears to work properly (see `takeTBMQueueWithTimeout` and `chunkSource`). Perhaps there are still bugs in it which I have not observed, though.
2. Previously pretty much everything was plumbed into `MonadError`; this caused a lot of bureaucracy and made things a lot more complicated & ornate than they needed to be. I've made the thing internally use normal exceptions, but exposed APIs returning a suitable error type.
3. Shutdown is cleaner now. When the scope of continuation passed to `withKinesisProducer` exits, then the remaining messages are flushed out with a configurable timeout.
4. A lot of internal stuff was using `MonadBaseControl IO`, etc., but I found that no expressivity had been gained over just using `IO` directly. So a number of things have been simplified in that respect, though the public APIs are all polymorphic over their monad.

Some things to look for in particular:
1. Places where exceptions need to be caught but perhaps aren't; or things that need to be masked
2. Policies for respawning things... For instance, what happens if the producer worker dies? Should it be restarted? Should this logic be handled here or in the client code?

larskuhtz commented:
I'll take a look. I'll probably need some time to read through all of it.

> Previously pretty much everything was plumbed into MonadError; this caused a lot of bureaucracy and made things a lot more complicated & ornate than they needed to be. I've made the thing internally use normal exceptions, but exposed APIs returning a suitable error type.

While working on similar issues in yet-another-logger I just made a few experiences that make me wonder if the apparent complexity of `MonadError` and the apparent simplicity of normal Exceptions is somewhat spurious. Maybe with `MonadError` certain edge cases are just explicit and one is forced to deal with them, while with normal exceptions it's easy to overlook certain subtleties in the semantics of throw and catch. The more I dig in the exceptions in a multithreaded setup the more I wonder if I really understand how they work.

jonsterling committed:
[Producer] if worker dies, wrap its exception

jonsterling committed:
[Producer] fix infinite loop in takeTBMQueueWithTimeout

jonsterling committed:
[Producer] better termination logic for chunkedSourceTBMQueue

terminate only if the queue is closed & empty

jonsterling committed:
[Producer] simplify chunking code & zealously close queues

jonsterling committed:
Minor cleanup

- log respawn of worker
- remove redundant imports
- remove dependency on stm-conduit

larskuhtz commented on cli/CLI.hs:
```diff
@@ -99,7 +95,7 @@ fetchCredentials = do
     True →
       loadCredentialsFromInstanceMetadata
-        <!?> KinesisError (toException NoInstanceMetadataCredentials)
+        ≫= maybe (throw NoInstanceMetadataCredentials) return
```
What about using `throwIO` instead of `throw`?

jonsterling replied:
Good point! I think I made that mistake in a number of places.

jonsterling committed:
Replace a bunch of 'throw' with 'throwIO'

Thanks @larskuhtz for catching this! It's a typo that I make constantly.

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+
+  let
+    reshardingLoop =
+      forever ∘ handle (\(SomeException _) → threadDelay 3000000) $ do
```
I am always a little concerned when I see this kind of loops, in particular in a thread. If there is chance that the system goes into a bad state, this could loop forever without doing reasonable work while the reset of the system is working under the assumption that this component does it's job. It seems that `updateStreamState` is doing network IO, which could be vulnerable to such an situation. For instance the name resolution may get into a state of permanent failure. 

If this would happen, is there a mechanism that would notify the overall system and trigger appropriate actions?

larskuhtz replied:
Is it guaranteed that `reshardingLoop` can't exit due to an asynchronous exception? Since it's a local let binding and only used right below in a thread for which the `ThreadID` is not exposed to other code, it's probably safe. Otherwise, I think, that thread should be spawned in with exceptions masked and only the inner action of `handle` should be executed umasked.

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+        threadDelay $
+          case recordsCount of
+            0 → 5000000
+            _ → 70000
```
Maybe you could an comment why you choose the value for above constants? I guess the values somehow related to properties of the AWS Kinesis API.

jonsterling replied:
Right; very sadly, these constants are the result of many weeks of tweaking in order to avoid rate limiting errors. 
- [ ] I'll add a comment to that effect.

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+        threadDelay 10000000
+
+    producerLoop =
+      forever ∘ handle (\(SomeException _) → threadDelay 2000000) $ do
```
Same comments as for the previous function apply.

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+            _ → 70000
+
+  withAsync (liftIO reshardingLoop) $ \reshardingHandle → do
+    link reshardingHandle
```
From the code above it seems to me that `reshardingHandle` shouldn't be able to throw exceptions. What would be the purpose of link in that case?

jonsterling replied:
You are quite right; I think it was there to anticipate when I should add some kind of a limit to the retrying of the workers... Probably fine to remove it at this point.

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+  withAsync (liftIO reshardingLoop) $ \reshardingHandle → do
+    link reshardingHandle
+    withAsync (liftIO producerLoop) $ \producerHandle → do
+      link producerHandle
```
Same here, what is the purpose of `link`?

jonsterling committed:
[Consumer] add comment explaining magic constants

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+      iteratorVar ← newTVarIO $ Just it
+      sequenceNumberVar ← newTVarIO Nothing
+      return ShardState
+        { _ssIterator = iteratorVar
```
This pattern looks interesting. I try to understand why a new `TVar` is used here instead of just updating the existing var. 
- Is the content of the reference updated somewhere in the code? Otherwise maybe a immutable value could be used instead?
- Are there users of the old reference after this field got updated with a new reference?

jonsterling replied:
What existing var?

jonsterling replied:
By the way @larskuhtz, none of this code is new—whilst it could certainly benefit from improvement, the thing that would benefit from the most focus would be the Producer.

larskuhtz replied:
I mean the `iteratorVar` reference in the StreamState that is given as argument.

jonsterling replied:
Oh I see. Right now, no attempt is made to match the shards returned from `GetStreamDescription` to the ones we already have. It would be good to consider doing so though.

jonsterling replied:
(To be clear, any shards returned from `GetStreamDescription` which are not "new" will be discarded, if I recall correctly. But it has been months since I touched this code, so take what I say with a grain of salt.)

larskuhtz commented on src/Aws/Kinesis/Client/Consumer.hs:
```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
+      return ShardState
+        { _ssIterator = iteratorVar
+        , _ssShardId = shardShardId
+        , _ssLastSequenceNumber = sequenceNumberVar
```
Same comment as for `_ssIterator`.

larskuhtz commented on src/Aws/Kinesis/Client/Producer.hs:
```diff
@@ -354,111 +318,101 @@ generatePartitionKey gen =
+  let
+    timeoutAction = do
+      threadDelay timeout
+      atomically $ putTMVar timedOutVar ()
```
Could you instead use `registerDelay` from `Control.Concurrent.STM.TVar`?

larskuhtz commented on src/Aws/Kinesis/Client/Producer.hs:
```diff
@@ -354,111 +318,101 @@ generatePartitionKey gen =
+
+  withAsync timeoutAction $ \h → do
+    xs ← atomically $ go []
+    xs <$ cancel h
```
I wonder if with `registerDelay` all of this could be coded within a single `atomically`. Not sure if would make sense, though.

jonsterling replied:
Interesting idea! I did not know about `registerDelay`, but it looks like it could be useful here.

jonsterling replied:
Actually, after thinking a bit about what it would mean to put it into a single `atomically`, I wonder if the current code is even correct as-is.

The thinking was that the `readTMVar` would retry until it the timeout fired, but this would actually end up retrying the entire transaction as a whole, so that doesn't seem right. Do you have any thoughts or clarifications?

larskuhtz commented on src/Aws/Kinesis/Client/Producer.hs:
```diff
@@ -354,111 +318,101 @@ generatePartitionKey gen =
+      (&&) <$> isClosedTBMQueue q <*> isEmptyTBMQueue q
+
+  unless shouldTerminate $ do
+    items ← lift $ takeTBMQueueWithTimeout _cpMaxChunkSize _cpMinChunkingInterval q
```
You could consider inlining `shouldTerminate` into `takeTBMQueueWithTimeout` by just using `readTBMQeue` which retries as long as the queue is not closed _and_ empty.

jonsterling replied:
I think it would not be much of a gain here (unless I missed something), since the purpose here is to determine whether `chunkedSourceTBMQueue` should recurse, not whether an STM transaction should retry.

But perhaps I have misunderstood...

larskuhtz replied:
There are two reasons why

``` haskell
items ← lift $ takeTBMQueueWithTimeout _cpMaxChunkSize _cpMinChunkingInterval q
```

could possibly return `[]`:
1.  the queue is closed or
2.  there was a timeout.

If there was a timeout the code recurses immediately. I wonder if `takeTBMQueueWithTimeout` could be changed to simply not return but retry when an timeout occurs before any items are received. In that case the definition of `shouldTerminate` and the guard `unless shouldTerminate` could be omitted. The function `go` above would be changed to contain:

``` haskell
readItems xs <|> if null xs then retry else xs <$ timeOut
```

The semantics of `go` would be slightly more complex (but a short comment should still be enough to explain it) while the rest of the code would become simpler.

``` haskell
chunkedSourceTBMQueue cp@ChunkingPolicy{..} q = do
  items ← lift $ takeTBMQueueWithTimeout _cpMaxChunkSize _cpMinChunkingInterval q
  unless null items $ do
    yield items
    chunkedSourceTBMQueue cp q
```

jonsterling replied:
Very interesting, thank you for clarifying this. I'll play around with your suggestions and see if I can come up with something that works nicely in the sense you describe.

jonsterling replied:
I have thought some more about this approach, and realized that whilst it would simplify the code of `chunkedSourceTBMQueue`, it may also make the semantics of `takeTBMQueueWithTimeout` more confusing. How does one justify a function like this that purports to have a timeout blocking indefinitely on an empty queue?

larskuhtz replied:
The semantics would be the same as what many handle/streaming frameworks use for synchronous calls: It either returns some data (subject to the requested size and timeout) or it returns no data, where latter only happens when the stream is closed.

jonsterling replied:
You are probably right; I think in this case, though, I had better call it something other than `takeTBMQueueWithTimeout`, since `withTimeout` seems to indicate that it will stop blocking after a timeout, but there are situations under which this will not be the case.

jonsterling replied:
I thought about it even more, and I think that the proposed semantics are precisely what we do not want for this function.

Imagine if there are five items in the queue, and we have requested ten. The whole point of the chunking mechanism here is to make sure that we keep yielding items at _at least_ the specified interval/timeout; it may be that some chunks have fewer items than we requested, but this is desired. The purpose is to avoid a situation where certain messages never make it through to Kinesis, because we stopped writing to the queue before we hit the proper chunk-size.

larskuhtz replied:
I think you miss in your argument that the timeout is still effective when there are more than 0 items. The semantics are:
- If there is no item at all block until an item arrives or the queue is closed (and ignore the timeout).
- if there is at least one item in the queue wait for either at most `n` items with timeout `t`.

One thing to consider it to reset the timeout when the queue was empty for longer than the timeout and the first item arrives. But I think it makes sense to just return one item in this case.

jonsterling replied:
I don't think those semantics hold for the proposed implementation. If the queue becomes empty in one of the recursive calls to `go`, the desired behavior is: **wait for the timeout and then return the items we got**. The actual behavior would be, **block until more items get added to the queue**. Unless I am misunderstanding something...

larskuhtz replied:
OK, here is the proposed implementation:

``` haskell
let
  -- if the queue is closed, then return what we have already got;
  -- otherwise, retry until we can read an item from it.
  readItems xs =
    readTBMQueue q ≫= \case
      Nothing → return xs
      Just x → go (x:xs)

  -- retry until the timeout fires
  timeout = readTVar timedOutVar ≫= check

  go xs
    -- if we have got enough items, return immediately
    | length xs ≥ n = return xs
    -- either try reading from the queue, or, unless the queue is empty, 
    -- check we have run out of time
    | otherwise = readItems xs <|> if null xs then retry else xs <$ timeOut
```

In `go` first we try if we can read an item. If that fails (queue is neither closed or there is an item) we
try either
-   do nothing in case the we haven't yet read any item at all or
-   we check the timeout, if we have already read at least one item.

So, we block until more items get added to the queue only if
-   we haven't read any item at all,
-   we read some items and the timeout triggered (either before or after reading the items), or
-   we read more as much items as requested.

This means the only case an empty list is returned is when the queue is closed. Otherwise an non-empty list is returned precisely if either the timeout has triggered (possibly in the past) or we got as many items as requested.

I think that's the desired behavior.

larskuhtz replied:
Note that I substituted the word _block_ with _retry_ in the code-comments in the previous comment. I think that explains better the semantics of the `Alternative` instance for `STM`.

jonsterling replied:
> - we read some items and the timeout triggered (either before or after reading the items), or

Can you clarify this case? If the timeout has triggered, then we want to return what items we at least have read, rather than ~~block~~ retry, right?

larskuhtz replied:
When the timeout has triggered we returns whatever items we have read until we have read at least one item or the queue is closed.

That means if the timeout triggers
-   we return immediately all items, if the list of items is non-empty,
-   we return the first item to arrive, if the list of items is empty,
-   we return the empty list, if the queue is closed before any items arrives.

larskuhtz replied:
I think that's a common semantics for synchronous IO system calls:

> return at least one (but possibly more) items from the stream subject to the buffer size and a timeout.

larskuhtz replied:
I think that's a common semantics for synchronous IO system calls:

> return at least one (but possibly more) items from the stream subject to the buffer size and a timeout.

larskuhtz replied:
Well, actually the "subject to a timeout part" is probably not standard semantics.

jonsterling replied:
> That means if the timeout triggers
> - we return immediately all items, if the list of items is non-empty,
> - we return the first item to arrive, if the list of items is empty,
> - we return the empty list, if the queue is closed before any items arrives.

I agree with all that; that seems proper. I just see the following statement in your previous comment that concerns me: 

> So, we block until more items get added to the queue only if we read some items and the timeout triggered (either before or after reading the items)

That does not seem consistent with the first bullet point of your most recent comment.

larskuhtz replied:
I think the confusing point in this semantics is that the timeout only becomes effective if at least one item has been read. That's effectively the same as the old behavior where the calling function would recurse immediately when an empty list is returned due to an timeout. So the new implementation would safe those explicit calls and would probably be less busy. 

In the old implementation the retry is done explicitly by returning, testing, and calling again, in the new implementation the retry is done event driven by STM. But maybe that's too confusing and the more explicit (but busy) way to implement it is better.

larskuhtz replied:
hmm, you are right the second quote in your previous comment doesn't make sense. Seems that I was confused. I think I that I negated the premise after the first bullet point in the list of conclusions in the original place. Sorry for the confusion.

larskuhtz replied:
The following code would be a little more concise but I don't really know which version is easier to read.

``` haskell
  | otherwise = readItems xs <|> do 
      t <- readTVar timedOutVar
      if not t || null xs then retry else return xs
```

jonsterling committed:
[Producer] implement takeTBMQueueWithTimeout using registerDelay

larskuhtz commented on src/Aws/Kinesis/Client/Producer.hs:
```diff
@@ -354,111 +319,95 @@ generatePartitionKey gen =
+          res ← Left <$> readTBMQueue q <|> Right <$> (readTVar timedOutVar ≫= check)
+          case res of
+            Left mx → maybe (return xs) (go ∘ (:xs)) mx
+            Right () → return xs
```
I wonder if something like the following would be easier to read (and be potentially more efficient because it avoids packing and unpacking with `Either`, though that's probably optimized away anyways).

``` haskell
let 
    readItems xs = readTBMQueue q ≫= \case
        Nothing → return xs
        Just x → go (x:xs)
    timeOut = readTVar timeOutVar ≫= check
go xs
    | length xs ≥ n = return xs
    | otherwise = readItems xs <|> xs <$ timeOut
```

jonsterling replied:
Hmmm, that code does look at lot clearer. I think it is observationally equivalent as well; I'll try it out and probably switch to it. Thanks a lot for the tip!

jonsterling committed:
[Producer] refactor loop in takeTBMQueueWithTimeout

jonsterling committed:
[Producer] Add comments summarizing semantics of takeTBMQueueWithTimeout

jonsterling committed:
[Producer] add lens for _pkCleanupTimeout

jonsterling committed:
[Producer] cargo-cult an {-# UNPACK #-}

jonsterling committed:
[Producer] add makeProducerKit

jonsterling committed:
[Consumer] add makeConsumerKit

jonsterling committed:
[CLI] refactor construction of kinesis & consumer kits

larskuhtz commented on src/Aws/Kinesis/Client/Producer.hs:
```diff
@@ -354,111 +319,99 @@ generatePartitionKey gen =
+      | length xs ≥ n = return xs
+      | otherwise = readItems xs <|> xs <$ timeout
+
+  atomically $ go []
```
I think putting `go` into a single atomic transaction is not a good idea here (sorry for proposing it). It will wait until there are enough items in the queue or the timeout triggers, but it will do so in a very busy way, by counting the available items (starting from zero) each time a new item is added to the list. At least it won't interfer with the writers, since it's retrying most of the time, thus not causing any harm. However, since the final successful transaction is pretty large chances are the the writers interfer with the that successful read operation if it involves a role-over of the internal read and write lists.

I think the following code is better:

```
let
  readItems xs = atomically (readTBMQueue q) ≫= \case
    Nothing → return xs
    Just x → go (x:xs)

  go xs
    | length xs ≥ n = return xs
    | otherwise = do 
        t ← atomically $ readTVar timedOutVar
        if not t || null xs then readItems xs else return xs
```

jonsterling replied:
I played around with this code, and the following seems to happen: 
1. I put `1000` items in the queue, and I try to bite off `2000` items in `5s`.
2. `go` recurses until it has got `1000` items, and the timeout has not yet fired; then because `not t` is true, then `readItems` is called which blocks until another item is added to the queue. Meanwhile, the timeout fires, but nobody hears about it and we just block.

jonsterling replied:
Not to mention, it's really to get this code to throw `thread blocked indefinitely in STM transaction`.

larskuhtz replied:
You are right that's an issue. A solution could be include the timeout check into the atomically block. This is a pretty cheap check that shouldn't affect the odds of go to win the race on the queue too much.

larskuhtz replied:
Regarding the `thread blocked indefinitely in STM transaction` exception: you have to be careful in particular if this is latest operation in the program. This exception is triggered when a transaction is retrying but there is no other reference to any variable in the transaction. This could for instance be the case if the producer loop exists, but the consumer is not explicitly terminated. This case would be unrelated to the correctness of the inner consumer or producer loop.

jonsterling replied:
Yeah, I think that will be necessary. It also then necessitates our (hypothesized) queue structure to live in STM rather than IO.

Sent from my iPad

> On Mar 27, 2015, at 5:32 PM, Lars Kuhtz notifications@github.com wrote:
> 
> In src/Aws/Kinesis/Client/Producer.hs:
> 
> > -  timedOutVar ← registerDelay timeoutDelay
> >   +
> > -  let
> > -    readItems xs =
> > -      readTBMQueue q ≫= \case
> > -        Nothing → return xs
> > -        Just x → go (x:xs)
> >   +
> > -    timeout =
> > -      readTVar timedOutVar ≫= check
> >   +
> > -    go xs
> > -      | length xs ≥ n = return xs
> > -      | otherwise = readItems xs <|> xs <$ timeout
> >   +
> > -  atomically $ go []
> >   You are right that's an issue. A solution could be include the timeout check into the atomically block. This is a pretty cheap check that shouldn't affect the odds of go to win the race on the queue too much.
> 
> —
> Reply to this email directly or view it on GitHub.

jonsterling committed:
[Producer] Preliminary Queue abstraction

jonsterling committed:
[Producer] Update to use variant of Lars's queue interface

Add dummy implementation for TBMQueue.

jonsterling committed:
[Producer, Queue] use Natural type where it makes sense

All these things can never be negative.

jonsterling committed:
[Producer] Remove InvalidProducerKit exception, no longer needed

jonsterling committed:
[Producer] Fix reference to undefined symbol

jonsterling committed:
[Producer] fix & refactor clean shutdown

jonsterling committed:
[Producer] Remove support for PutRecord endpoint (close #22)

https://github.com/alephcloud/hs-aws-kinesis-client/issues/22

It makes no sense to use PutRecord for this kind of thing now that
everybody has access to PutRecords.

jonsterling committed:
[Producer] Consistent unicode use

jonsterling committed:
[Producer] Remove some unnecessary annotations

jonsterling committed:
Add a TBMChan queue implementation

jonsterling committed:
[Producer] Use existential to clean up queue parameterization

jonsterling committed:
[Producer] Put queue implemenation proxy into the Producer Kit

This way you can just omit it and the default one will be used.

jonsterling committed:
[Producer] Only use concrete Proxy type where needed

jonsterling committed:
[Producer] Slight modularization

jonsterling committed:
[Producer] Document QueueImplementation type

jonsterling committed:
Update changelog

jonsterling committed:
[Producer] Move queue chunking to its own module

jonsterling committed:
Clean up unused language extensions

jonsterling committed:
[CHANGELOG] word-smithing

jonsterling committed:
[Queue] implement takeQueueTimeout w/o bit atomically block

The previous version had the whole thing inside an `atomically` block,
which resulted in horrible performance, since the entire operation would
get retried all the time. This version worked sort of OK with `TQueue`,
but it could result in the system locking up for many seconds with
`TChan`.

This time, I have implemented it such that only the individual reads
from the queue are atomic; it should be functionally equivalent, if I am
not mistaken.

jonsterling committed:
[Consumer] use Natural rather than Int for batch size

jonsterling committed:
Update constraints file

I don't believe in freezing, but it is nice to *know* a set of
constraints at which the package worked successfully.

jonsterling committed:
[Consumer] some modularization

jonsterling committed:
Fix copyright dates to 2015

jonsterling committed:
[Consumer] don't both to qualify Aws.Kinesis import

jonsterling committed:
[Consumer] move out a few more internals

jonsterling committed:
[Consumer] remove redundant language extensions

jonsterling committed:
[Producer] change cleanup timeout to be microseconds

Seems pretty standard for things like this

jonsterling committed:
Add debug logging via cabal flag

Resolves #21

jonsterling committed:
Unindent preprocessor directives; fix warnings

- some versions of CPP do not work with indented directives
- made a few imports conditional to avoid warnings when -fdebug is turned off

jonsterling commented:
I am going to merge this now. To summarize:
1. The Producer is much improved in multiple respects. When its scope exits, it cleanly shuts down (flushing messages with a timeout); the old behavior was that the producer would never exit until its owning thread was killed.
2. There is a bug in STM's `TQueue` which causes live-locking under heavy load; the bug was a strict reverse of the internal queue, which should have been lazy. The Producer now ships by default using a queue based on `TBMChan`, but it may be configured to use `TBMQueue` (or any other implementation), which user may wish to do once STM is patched, since `TQueue` has better throughput in general than `TChan`.
3. Many things have been cleaned up & modularized. Exception handling has been adjusted such that exceptions are used internally where it makes sense, and APIs which expect you to handle errors are exposed returning `m (Either SpecificErrorType a)`.
4. The Consumer is not much changed; because it has not had the benefit of much focused work, there are still problems with it, but they date back to before this version. In other words, this version does not make the Consumer any _worse_.
